tutorial/tests/test_functions_advanced.py

```diff
@@ -0,0 +1,259 @@
+import pathlib
+import time
+import typing as t
+from operator import sub
+from string import ascii_lowercase as lowercase
+from string import ascii_uppercase as uppercase
+from string import digits, punctuation
+
+import pytest
+
+
+#
+# Exercise: Password checker factory
+#
+def reference_password_checker_factory(
+    min_up: int, min_low: int, min_pun: int, min_dig: int
+) -> t.Callable:
+    """Password checker factory"""
+
+    def password_checker(password: str) -> tuple[bool, dict]:
+        """Password checker function"""
+
+        # Counts the number of chars for each class in a password
+        counts = [
+            sum(1 for char in password if char in _class)
+            for _class in (uppercase, lowercase, punctuation, digits)
+        ]
+
+        # Compare with requirements and calculate the differences
+        diffs = [
+            sub(*pair) for pair in zip(counts, (min_up, min_low, min_pun, min_dig))
+        ]
+
+        result = dict(zip(("uppercase", "lowercase", "punctuation", "digits"), diffs))
+
+        return all(diff >= 0 for diff in diffs), result
+
+    return password_checker
```

Comment:
Consider alwo the following solution:

```python
def solution_password_checker_factory(min_up: int, min_low: int, min_pun: int, min_dig: int):
    """Password checker factory"""
    def password_checker(pwd: str):
        
        def n_elements(method):
            return len([l for l in pwd if getattr(l, method)()])
                
        return_dict = {
            'uppercase': n_elements("isupper") - min_up,
            'lowercase': n_elements("islower") - min_low,
            'punctuation': len(pwd) - n_elements("isalnum") - min_pun,
            'digits': n_elements("isdigit") - min_dig,
        }

        flag = all([v >= 0 for v in return_dict.values()])
        return flag, return_dict
    
    return password_checker
```

The advantage here is that when revealing the reference, no surprising variables appear from nowhere (like  `uppercase`, `lowercase`, etc.).

Alternatively, you can maybe import them directly in the function.


Up to you, just a suggestion.

Reply:
I like your suggestion for the solution, but I need the `from string import ...` because the tests won't work otherwise as I'm accessing these names from the `globals()` dictionary.